experimental/source/source.go

```diff
@@ -45,14 +39,36 @@ type Opener interface {
 	Open(path string) (string, error)
 }
 
-// Map implements [Opener] via map lookup.
+// Map implements [Opener] via lookup of a built-in map. This map is not
+// directly accessible, to help avoid mistaken uses that cause different *Map
+// pointer values (for the same built-in map value) to wind up in different
+// queries, which breaks query caching.
 //
 // Missing entries result in [fs.ErrNotExist].
-type Map map[string]string
+type Map struct {
+	table map[string]string
+}
+
+// NewMap creates a new [Map] wrapping the given map.
+//
+// If passed nil, this will update the map to be an empty non-nil map.
```

Comment:
Why? Since it's the `*Map` pointer value that is comparable and determines identity, there doesn't seem to be any value in canonicalizing the map field this way.

Reply:
Because if a user writes `NewMap(nil)`, `Get()` can return `nil`. I'd like to avoid reproducing this minor footgun from Go maps into my own type.

Reply:
Why is that a foot gun? A `nil` map is a reasonable value in Go and behaves like a read-only empty map.

Reply:
Er, I guess this is related to the "This can be used to modify the map" part of Go docs for `Get()`. But why is that an expected/supported way to modify the map? I guess I would imagine that the caller instead would retain a reference to the map provided to `NewMap`, and use that if they needed to mutate it.